utils/artefacts/artefacts.go

```diff
@@ -0,0 +1,148 @@
+)
+
+type (
+	getArtefactManagerFunc = func(ctx context.Context, job, artefact string) (*client.ArtefactManagerItem, *http.Response, error)
+	getOutputArtefactFunc  = func(ctx context.Context, job, artefactID string) (*os.File, *http.Response, error)
+)
+
+type ArtefactManager struct {
+	getArtefactManager getArtefactManagerFunc
+	getOutputArtefact  getOutputArtefactFunc
+}
+
+func NewArtefactManager(getArtefactManager getArtefactManagerFunc, getOutputArtefact getOutputArtefactFunc) *ArtefactManager {
+	return &ArtefactManager{
+		getArtefactManager: getArtefactManager,
+		getOutputArtefact:  getOutputArtefact,
+	}
+}
+
+func (m *ArtefactManager) DownloadJobArtefact(ctx context.Context, jobName string, outputDirectory string, artefactManagerItem client.HalLinkData) (err error) {
+	artefactManagerName := artefactManagerItem.GetName()
+	artefactManager, resp, apierr := m.getArtefactManager(ctx, jobName, artefactManagerName)
+	defer func() {
+		if resp != nil {
+			_ = resp.Body.Close()
+		}
+	}()
+	err = utils.CheckAPICallSuccess(ctx, fmt.Sprintf("cannot fetch artefact manager [%v]", artefactManager), resp, apierr)
+	if err != nil {
+		return
+	}
+
+	artefactFilenamePtr, ok := artefactManager.GetTitleOk()
+	if !ok {
+		err = fmt.Errorf("%w: could not fetch artefact title from artefact manager [%v]", commonerrors.ErrUndefined, artefactManagerName)
+		return
+	}
+	artefactFilename := *artefactFilenamePtr
+
+	expectedSizePtr, ok := artefactManager.GetSizeOk()
+	if !ok {
+		err = fmt.Errorf("%w: could not fetch artefact size from artefact manager [%v]", commonerrors.ErrUndefined, artefactManagerName)
+		return
+	}
+	expectedSize := *expectedSizePtr
+
+	expectedHashPtr, ok := artefactManager.GetHashOk()
+	if !ok {
+		err = fmt.Errorf("%w: could not fetch artefact hash from artefact manager [%v]", commonerrors.ErrUndefined, artefactManagerName)
+		return
+	}
+	expectedHash := *expectedHashPtr
+
+	artefactIDPtr, ok := artefactManager.GetNameOk()
+	if !ok {
+		err = fmt.Errorf("%w: could not fetch artefact name from artefact manager [%v]", commonerrors.ErrUndefined, artefactManagerName)
+		return
+	}
+	artefactID := *artefactIDPtr
+
+	artefact, resp, apierr := m.getOutputArtefact(ctx, jobName, artefactID)
+	defer func() {
+		if resp != nil {
+			_ = resp.Body.Close()
+		}
+		if artefact != nil {
+			_ = artefact.Close()
+		}
+	}()
+
+	err = utils.CheckAPICallSuccess(ctx, fmt.Sprintf("cannot fetch generated artefact [%v]", artefactFilename), resp, apierr)
+	if err != nil {
+		return
+	}
+
+	unescapedName, err := url.PathUnescape(artefactFilename)
+	if err == nil {
+		artefactFilename = unescapedName
+	}
+
+	destinationPath := filepath.Join(outputDirectory, artefactFilename)
+	destination, err := filesystem.CreateFile(destinationPath)
+	if err != nil {
+		err = fmt.Errorf("%w: could not create a location to store generated artefact [%v]: %v", commonerrors.ErrUnexpected, artefactFilename, err.Error())
```

Comment:
this helper can be used https://github.com/ARM-software/golang-utils/blob/519c2fe8b29c57bec091b7cb357570ebb091c1d9/utils/filesystem/files.go#L408